### PR TITLE
Fix #2080

### LINKED ISSUE #2080: beatport: Fetch musical key, BPM, and genre data
### Problem

Everything seems to be getting pulled in except BPM & Key. Genre tag appears to be hit or miss. All tracks are from beatport, so the tags should be there. Looking at the API docs I see the facets listed. Are these actually accessible in Track Info?

```
Requires catalog API Permission 

Content can be filtered on facets. Ex:facets=genreName:House, subGenreName:Hard 

Use returnFacets to obtain specific information about the return set.

Allowed facets/returnFacets:

Note: Not all facets are relevant to all content types.

artistName
performerName
artist
artistId
performerId
label
labelName
labelId
subGenre
subGenreName
subGenreId
genre
genreName
genreId
bpm         <------------
key          <------------
tagId
```
### Setup
- OS: Mac OS X 10.11.5 (15F34)
- Python version: 2,7
- beets version: 1.3.18
- Turning off plugins made problem go away (yes/no): no

My configuration (output of `beet config`) is:

``` yaml
briansmac:beets anodigital$ beet config
lyrics:
    bing_lang_from: []
    bing_client_secret: REDACTED
    google_API_key: REDACTED
    force: no
    google_engine_ID: REDACTED
    sources:
    - google
    - lyricwiki
    - lyrics.com
    - musixmatch
    auto: yes
    fallback:
    genius_api_key: REDACTED
    bing_lang_to:
library: /Users/anodigital/Desktop/music.blb
original_date: yes
embedart:
    maxwidth: 1000
    compare_threshold: 0
    auto: yes
    ifempty: no
    remove_art_file: no
pluginpath: ~/github/beets/beetsplug

plugins: beatport fromfilename discogs embedart fetchart ftintitle fuzzy importadded lyrics mbsync spotify types
beatport:
    source_weight: 0.5
    tokenfile: beatport_token.json
    apikey: REDACTED
    apisecret: REDACTED
threaded: yes
art_filename: cover
duplicate_action: remove
none_rec_action: skip
ignore: .AppleDouble ._* *~ .DS_Store
directory: /Users/anodigital/Desktop/Music

import:
    copy: no
    write: yes
    resume: yes
    quiet_fallback: skip
    timid: no
    log: beetslog.txt
group_albums: no

match:
    strong_rec_thresh: 0.05
    medium_rec_thresh: 0.25
    rec_gap_thresh: 0.25
    max_rec:
        missing_tracks: medium
        unmatched_tracks: medium
    ignored: missing_tracks unmatched_tracks
discogs:
    source_weight: 0.5
    tokenfile: discogs_token.json
    apikey: REDACTED
    apisecret: REDACTED
fuzzy:
    threshold: 0.7
    prefix: '~'
spotify:
    regex: []
    artist_field: albumartist
    mode: list
    track_field: title
    album_field: album
    region_filter:
    show_failures: no
    tiebreak: popularity
importadded:
    preserve_write_mtimes: no
    preserve_mtimes: no
fetchart:
    auto: yes
    minwidth: 0
    sources:
    - filesystem
    - coverart
    - itunes
    - amazon
    - albumart
    google_engine: 001442825323518660753:hrh5ch1gjzm
    enforce_ratio: no
    cautious: no
    maxwidth: 0
    store_source: no
    google_key: REDACTED
    fanarttv_key: REDACTED
    cover_names:
    - cover
    - front
    - art
    - album
    - folder
ftintitle:
    auto: yes
    drop: no
    format: feat. {0}
```

### PR DESCRIPTION
**Pull request by temrix**
This fixes feature request #2080 .

I have added `bpm`, `musical_key` and `genre` in the beatport plugin. I needed to modify the hooks in the autotagger to accept those additional attributes. `genre` will be fetched for each track. I have also added a test file to test the plugin.

### CODE REVIEW

File: docs/plugins/beatport.rst [hunk 1]
```diff
@@ -21,15 +23,18 @@ run the :ref:`import-cmd` command after enabling the plugin, it will ask you
 to authorize with Beatport by visiting the site in a browser. On the site
 you will be asked to enter your username and password to authorize beets
 to query the Beatport API. You will then be displayed with a single line of
-text that you should paste into your terminal. This will store the
-authentication data for subsequent runs and you will not be required to
-repeat the above steps.
+text that you should paste as a whole into your terminal. This will store the
+authentication data (under ``~/.config/beatport_token.json``) for subsequent
+runs and you will not be required to repeat the above steps.
```
Comment (sampsyo) [inline docs/plugins/beatport.rst]: The path to the token file is actually configurable, so I think we should avoid mentioning a hard-coded path here. (And it doesn't seem like most users need to worry about where this is stored?)

Comment (temrix) [reply docs/plugins/beatport.rst]: I agree.

**Comment by sampsyo** [comment]
The tests revealed that we might need to add `requests_oauthlib` to the test dependencies:
https://travis-ci.org/beetbox/beets/jobs/587184180#L930

Those are listed in our Tox config:
https://github.com/beetbox/beets/blob/3ffd5ca0c4a417e46bc60f61192a7051b85066b4/tox.ini#L13-L26

**Comment by temrix** [comment]
Ah, I have thought before that the `extras_require` in `setup.py` would take care of this but I did not know where I should put it instead.

**Comment by sampsyo** [comment]
Perfect! Let me re-emphasize that we _really_ appreciate your adding such extensive tests for the plugin. Wahoo!

**Comment by wualex** [comment]
@temrix i just updated my beets install and i tried to import a test dir with few tracks and i get the following error:

`Traceback (most recent call last):
  File "/usr/local/bin/beet", line 11, in <module>
    load_entry_point('beets==1.5.0', 'console_scripts', 'beet')()
  File "/usr/local/lib/python2.7/dist-packages/beets/ui/__init__.py", line 1267, in main
    _raw_main(args)
  File "/usr/local/lib/python2.7/dist-packages/beets/ui/__init__.py", line 1254, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/usr/local/lib/python2.7/dist-packages/beets/ui/commands.py", line 956, in import_func
    import_files(lib, paths, query)
  File "/usr/local/lib/python2.7/dist-packages/beets/ui/commands.py", line 926, in import_files
    session.run()
  File "/usr/local/lib/python2.7/dist-packages/beets/importer.py", line 329, in run
    pl.run_parallel(QUEUE_SIZE)
  File "/usr/local/lib/python2.7/dist-packages/beets/util/pipeline.py", line 445, in run_parallel
    six.reraise(exc_info[0], exc_info[1], exc_info[2])
  File "/usr/local/lib/python2.7/dist-packages/beets/util/pipeline.py", line 312, in run
    out = self.coro.send(msg)
  File "/usr/local/lib/python2.7/dist-packages/beets/util/pipeline.py", line 194, in coro
    func(*(args + (task,)))
  File "/usr/local/lib/python2.7/dist-packages/beets/importer.py", line 1353, in lookup_candidates
    task.lookup_candidates()
  File "/usr/local/lib/python2.7/dist-packages/beets/importer.py", line 641, in lookup_candidates
    autotag.tag_album(self.items, search_ids=self.search_ids)
  File "/usr/local/lib/python2.7/dist-packages/beets/autotag/match.py", line 460, in tag_album
    va_likely):
  File "/usr/local/lib/python2.7/dist-packages/beets/plugins.py", line 576, in decorated
    for v in generator(*args, **kwargs):
  File "/usr/local/lib/python2.7/dist-packages/beets/autotag/hooks.py", line 637, in album_candidates
    for candidate in plugins.candidates(items, artist, album, va_likely):
  File "/usr/local/lib/python2.7/dist-packages/beets/plugins.py", line 386, in candidates
    for candidate in plugin.candidates(items, artist, album, va_likely):
  File "/usr/local/lib/python2.7/dist-packages/beetsplug/beatport.py", line 360, in candidates
    return self._get_releases(query)
  File "/usr/local/lib/python2.7/dist-packages/beetsplug/beatport.py", line 412, in _get_releases
    for x in self.client.search(query)]
  File "/usr/local/lib/python2.7/dist-packages/beetsplug/beatport.py", line 138, in search
    release = self.get_release(item['id'])
  File "/usr/local/lib/python2.7/dist-packages/beetsplug/beatport.py", line 154, in get_release
    release.tracks = self.get_release_tracks(beatport_id)
  File "/usr/local/lib/python2.7/dist-packages/beetsplug/beatport.py", line 166, in get_release_tracks
    return [BeatportTrack(t) for t in response]
  File "/usr/local/lib/python2.7/dist-packages/beetsplug/beatport.py", line 265, in __init__
    self.genre = six.text_type(data['subGenres'][0].get('name'))
IndexError: list index out of range`

**Comment by temrix** [comment]
Can you open a new issue with this? We can discuss there. And try to pin-point the track or at least a range of albums the error has occurred with.

**Pull request merged**